### PR TITLE
KTX Compare Tool - Phase 2

### PR DESCRIPTION
**Pull request by aqnuep**
This now includes the complete implementation for the compare tool, as of the baseline, but once this is merged to the ktx-compare branch and that branch will result in a rebased PR against main, then I expect some minor changes.

This change also solves the issues that resulted in invalid golden files, because some of the compare tool testing depended on fixing the cause of those.

### CODE REVIEW

**Comment by aqnuep** [comment]
@MarkCallow, please review this change indifferent of the fact that it is a draft and CI may not pass yet.

This will anyway first go into the `ktx-compare` branch for finalization, rebasing, etc. before the final PR to main.

Thanks!

File: tools/ktx/command_create.cpp [hunk 1]
```diff
@@ -540,6 +540,18 @@ struct OptionsCreate {
...
         }
 
+        if (isFormatNotSRGBButHasSRGBVariant(vkFormat)) {
+            if (!convertOETF.has_value() && assignOETF.has_value() && assignOETF == KHR_DF_TRANSFER_SRGB) {
+                report.fatal_usage(
+                        "Invalid value to --assign-oetf \"{}\" for format \"{}\". Transfer function must not be sRGB for a non-sRGB VkFormat with sRGB variant.",
+                        args[kAssignOetf].as<std::string>(), args[kFormat].as<std::string>());
+            } else if (convertOETF.has_value() && convertOETF == KHR_DF_TRANSFER_SRGB) {
+                report.fatal_usage(
+                        "Invalid value to --convert-oetf \"{}\" for format \"{}\". Transfer function must not be sRGB for a non-sRGB VkFormat with sRGB variant.",
```
Comment (MarkCallow) [inline tools/ktx/command_create.cpp:550]: Can we not duplicate these long strings?

**Review by MarkCallow — CHANGES_REQUESTED** [review]
Well-spotted on the enum bitfields. Which compiler(s) had problems with them? I thought enums were usually ints under the covers so I'm surprised those definitions were not cross platform.

I spotted a couple of small things I'd like  you to change.

File: tools/imageio/imageio_utility.h [hunk 2]
```diff
@@ -122,6 +123,19 @@ inline uint16_t float_to_half(float value) {
 
 // -----------------------------------------------------------------------------
 
+inline int32_t sign_extend(uint32_t value, uint32_t numBits) {
+    assert(numBits <= 32);
+    uint32_t signBitMask = 1u << (numBits - 1);
+    if (value & signBitMask) {
+        uint32_t signExtendedValue = value | ~(signBitMask - 1);
+        int32_t result;
+        std::memcpy(&result, &signExtendedValue, sizeof(result));
```
[the remark this answers is not in the document]
Comment (aqnuep) [reply tools/imageio/imageio_utility.h:132]: memcpy guarantees bitwise copy, whereas other methods may hit undefined behavior land, hence generally these case are handled as such in most places. For small data this should be optimized out anyway.

File: tools/imageio/formatdesc.h [hunk 3]
```diff
@@ -550,7 +550,7 @@ struct FormatDescriptor {
         basic.primaries = p;
     }
     void setTransfer(khr_df_transfer_e t) {
-        khr_df_transfer_e oldOetf = basic.transfer;
+        khr_df_transfer_e oldOetf = static_cast<khr_df_transfer_e>(basic.transfer);
```
[the remark this answers is not in the document]
Comment (aqnuep) [reply tools/imageio/formatdesc.h]: Refactored.

File: tools/ktx/command_create.cpp (hunk 1 above)
Comment (aqnuep) [reply tools/ktx/command_create.cpp]: Refactored.

File: tools/imageio/imagespan.hpp [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+// vi: set sw=2 ts=4 expandtab:
+
+// Copyright 2023-2024 The Khronos Group Inc.
+// Copyright 2023-2024 RasterGrid Kft.
+// SPDX-License-Identifier: Apache-2.0
+
+//!
+//! @internal
+//! @~English
+//! @file imageiter.hpp
```
[the remark this answers is not in the document]
Comment (aqnuep) [reply tools/imageio/imagespan.hpp]: Fixed.

**Comment by aqnuep** [comment]
Hi @MarkCallow! Do you need more time to review this or the related CTS change or should I proceed with merging these into the ktx-compare development branch?

**Comment by MarkCallow** [comment]
> Do you need more time to review this or the related CTS change

I just attempted a review. Please see my comment over there.

**Comment by MarkCallow** [comment]
> > Do you need more time to review this or the related CTS change
> 
> I just attempted a review. Please see my comment over there.

The changes in this PR are fine. I am in the midst of a very busy period. I won't have time to start trying `ktx compare` until the middle of next week. I don't want to hold you up any more so go ahead and merge the CTS changes and this.

**Comment by aqnuep** [comment]
Thank you, @MarkCallow. I'll get everything merged in the ktx-compare development branch and then we'll start rebasing/integrating mainline chaines into that development branch, because that'll take some time too. After that I still intend to open PRs before merging the development branch into main, so you'll have the chance to do another review round and testing, as necessary.

**Comment by MarkCallow** [comment]
> I'm getting this error message:
> 
> ...

This is a recurring issue with Chocolatey. It is one of the reasons CI tries 4 times to download Doxygen before failing. I do not fully understand it and have not had much help from GitHub. I just tried re-running but it failed again. I suggest trying to re-run tomorrow.

**Pull request merged**